### DISCUSSION TITLE
DataTemplate binding never fires

### DISCUSSION THREAD

**Discussion by Varmod** [AvaloniaUI/Avalonia#12905] [Q&A] [answered]
**Describe the bug**
Context: I was trying to make a DataGridColumn with a checkbox that would update my data on click since I wasn't able to get the DataGridCheckBoxColumn to work the way I wanted.

The problem is that without the (sys:Boolean) cast (on a already boolean value) the code would throw the error:
`'Unable to cast object of type 'System.Boolean' to type 'Project.ViewModels.WindowViewModels.MyData`
However with the cast it compiles and I get a checkbox column, but the MyData.IsSelected Getter isn't hit when loading, or the Setter isn't hit when clicking the checkbox.

```xaml
<DataGrid x:DataType="s:MyData">
  <DataGrid.ItemsSource>
    <Binding Path="MyDataSource" x:DataType="vm:ViewModel" />
  </DataGrid.ItemsSource>
<DataGrid.Columns>
  <DataGridTemplateColumn>
    <DataGridTemplateColumn.CellTemplate>
      <DataTemplate>
        <ContentControl x:DataType="s:MyData" Content="{Binding Path=IsSelected}">
          <ContentControl.DataTemplates>
            <DataTemplate x:DataType="sys:Boolean">
              <!-- Problem with IsChecked binding -->
              <CheckBox x:DataType="s:MyData" IsChecked="{Binding Path=(sys:Boolean)IsSelected}" />
            </DataTemplate>
          </ContentControl.DataTemplates>
        </ContentControl>
      </DataTemplate>
    </DataGridTemplateColumn.CellTemplate>
  </DataGridTemplateColumn>
</DataGrid.Columns>
</DataGrid>
```

```cs
public class MyData
{
  public bool IsSelected { 
	  get; 
	  set; 
  }

  // More data that's bound in other columns
}
```

**Desktop (please complete the following information):**

- OS: Windows
- Version 11.0.4

**Additional context**
I found a working alternative: 
```xaml
<ContentControl x:DataType="s:MyData" Content="{Binding Path=IsSelected, Mode=TwoWay}" >
  <ContentControl.DataTemplates>
    <DataTemplate x:DataType="sys:Boolean">
      <CheckBox x:DataType="s:MyData" IsChecked="{Binding $parent[ContentControl].Content}"  />
    </DataTemplate>
  </ContentControl.DataTemplates>
</ContentControl>
```

**Comment by timunie** [comment, marked as answer]
Your DataTemplate refers to the wrong type actually. In DataGridTemplateColumn it is always the rows obejct you get, as it doesn't know about the field(s) your cell wants to use. 

That should work:

```xml
    <DataGridTemplateColumn.CellTemplate>
      <!-- Row's datatype goes here -->
      <DataTemplate DataType="s:MyData">
          <!-- No need to nest into another content control unless your are planning to add other stuff --> 
          <CheckBox IsChecked="{Binding Path=IsSelected}" />
      </DataTemplate>
    </DataGridTemplateColumn.CellTemplate>
```

**Reply by Varmod** [reply]
Looks like I overthought it. Your solution is what I was aiming for. Thanks

**Comment by timunie** [comment]
converting this into a discussion for now as I think it was just a typo. If it turns into an acutally bug, we can reopen this issue later.
